### PR TITLE
Hannah and Luxi - Scrabble - Octos

### PR DESCRIPTION
**Pull request by Lindseyls**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? | Git allows us to stay in-sync while making changes through our project. Also, gets us to frequently backup our work. 
Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | We made the 'score' method a class method so that we can easily access the 'score' method throughout the class.
Describe an example of a test you wrote for a _nominal case_. | We wrote a test for a nominal case in the test for Player#total_score method by instantiating the Player class and pass through words into the play(word) method. We needed to test that the total_score method return the expected total score. 
Describe an example of a test you wrote for an _edge case_. | We did not get a chance to do test for an edge case, however, where we could have put the test is in Player#draw_tiles method. What we would have done is given a draw number that is greater than 7 and have the test return nil. 
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We used the inject enumerable in the Player#total_score method. The inject method made it so that we did not have to pull the numbers out of the array, we just had to be able to manipulate the array to add up the scores. 
What was one method you and your pair used to debug code? | We used binding.pry to regularly check out methods. 
Is there a specific piece of code you'd like feedback on? | The WHOLE thing. :) 
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? |  Luxi - "I really enjoyed working with Hannah. I learned how to write and update test and using/relying on the stacktrace to point out the issues that we needed to change in order to pass the test. I also learned from Hannah to take it slow and not worry about what was coming in the class but to focus on the current method that we were working on." Hannah - "I really appreciated working with Luxi. We not only worked really well together but were also to ask questions of each other. We feel very comfortable sharing our ideas and approach to a problem and give regular feedback on how we should more forward. I also enjoyed learning more about pry from Luxi and having more experiencing using pry because that is not one of my preferred debugging methods."

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -1,9 +1,75 @@
...
+      elsif scrabble_word.length == 7
+        total_score += 50
+      end
+
+      # output the score of each letter in a word
+      scrabble_word.each do |letter|
+        case letter.downcase
+        when "a", "e", "i", "o", "u", "l", "n", "r", "s", "t"
+          total_score += 1
+        when "d", "g"
```
Comment (droberts-sea) [inline lib/scoring.rb]: Storing these as options in a `case` statement certainly works, but it is a single-use solution. If you ever needed to use letter values somewhere else, you'd have to copy paste. An alternative would be to build a data structure stored in a variable that maps letters to scores.

File: lib/scoring.rb [hunk 2]
```diff
@@ -1,9 +1,75 @@
...
+            top_word = [word]
+          elsif self.score(word) == top_score
+            top_word << word
+          end
+        end
+        if top_word.length == 1
+          return top_word[0]
+        elsif top_word.length > 1
+          if top_word[0] < top_word[1]
+            return top_word[0]
```
Comment (droberts-sea) [inline lib/scoring.rb]: The conditional on line 65 will choose the first word alphabetically. This doesn't match the tie-breaking logic we wanted. Moreover, this doesn't account for more than two tied words.

File: specs/scoring_spec.rb [hunk 3]
```diff
@@ -42,21 +43,32 @@
...
     end
 
     it 'returns the only word in a length-1 array' do
+      array_of_words = ["phone"]
+      Scrabble::Scoring.highest_score_from(array_of_words).must_be_same_as(array_of_words[0])
     end
 
     it 'returns the highest word if there are two words' do
+      array_of_words = ["dog", "frog"]
+      Scrabble::Scoring.highest_score_from(array_of_words).must_equal("frog")
```
Comment (droberts-sea) [inline specs/scoring_spec.rb]: It might be wise to reverse the order of this array and check that you get the same result. You would be surprised at how often this sort of thing turns up a bug. This applies to all the tiebreaker tests in this section.

File: lib/player.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      end
+      word_scores.inject(:+)
+    end # end of total_score
+
+    # this method takes the total score and see if the player won or not
+    def won?
+      if total_score > 100
+        return true
+      else
+        return false
```
Comment (droberts-sea) [inline lib/player.rb]: Since `>` will always return `true` or `false`, this could be shortened to:

```ruby
def won?
  return total_score > 100
end
```

File: specs/player_spec.rb [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+      word = "dog"
+      player_1 = Scrabble::Player.new("Hannah")
+      player_1.play(word)
+
+      player_1.plays.must_equal ["dog"]
+    end
+
+    it 'correctly returns false if player has already won' do
+      word_1 = "quetzal"
+      word_2 = "elevens"
```
Comment (droberts-sea) [inline specs/player_spec.rb:45]: What should it return if the player hasn't won yet?

File: specs/player_spec.rb [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+      player_1.play(word_4)
+
+      player_1.total_score.must_equal 26
+    end
+  end # end of describe 'total_score'
+
+  describe 'won?' do
+    it "correctly returns true if the player has over 100 points, otherwise returns false" do
+      word_1 = "quetzal"
+      word_2 = "elevens"
```
Comment (droberts-sea) [inline specs/player_spec.rb:75]: This only tests the case where it returns `true`. Other cases I'd be interested in seeing:

- What if the player has not played any words at all?
- What if the player has < 100 points?
- What if the player has exactly 100 points?

File: lib/tilebag.rb [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        count.times {@tiles << letter}
+      end
+    end # end of initialize
+
+    # this method returns a collection of random tiles
+    def draw_tiles(num)
+      tile_bag_count = 98
+      player_tiles = []
+      num.times do
+        tile_index = rand(0..tile_bag_count-1)
```
Comment (droberts-sea) [inline lib/tilebag.rb:54]: Since you hard code the `tile_bag_count` to 98, this may go past the end of the array if you draw tiles more than once.

File: specs/tilebag.spec.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    end
+  end # end of describe initialize
+
+  describe 'draw_tiles(num)' do
+    it "returns a collection of random tiles, removes the tiles from the default set" do
+      num = 3
+      draw_1 = Scrabble::Tilebag.new
+
+      draw_1.draw_tiles(num).must_be_kind_of Array
+      draw_1.draw_tiles(num).length.must_equal 3
```
Comment (droberts-sea) [inline specs/tilebag.spec.rb:33]: I'm not sure I like the variable name `draw_1` here - without reading line 27, I have no way of knowing what it is. Something like `tile_bag` might be a better choice.

File: specs/tilebag.spec.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+      end
+
+      tilebag.tiles.length.must_equal 98
+    end
+  end # end of describe initialize
+
+  describe 'draw_tiles(num)' do
+    it "returns a collection of random tiles, removes the tiles from the default set" do
+      num = 3
+      draw_1 = Scrabble::Tilebag.new
```
Comment (droberts-sea) [inline specs/tilebag.spec.rb:30]: Questions I'd like to see answered by your tests:

- Are the tiles actually removed from the bag?
- Am I able to draw the last tile from the bag?
- What happens if I try to draw more tiles than are currently in the bag?
- What if I try to draw 0 or -3 tiles?

File: lib/tilebag.rb [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+      "x" => 1,
+      "y" => 2,
+      "z" => 1
+    }
+
+    def initialize
+      @tiles = []
+      DEFAULT_TILES.each do |letter, count|
+        count.times {@tiles << letter}
+      end
```
Comment (droberts-sea) [inline lib/tilebag.rb:46]: I like this strategy of putting the right number of each letter into a big array. Much simpler than leaving it in that hash.

**Comment by droberts-sea** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | yes
Both Teammates contributed to the codebase | yes
Regular Commits with meaningful commit messages | yes
**Scoring class** |
`score` method | yes
Uses appropriate data structure to store the letter score | see inline
Appropriately handles edge cases | yes
`highest_score_from` method | yes
Appropriately handles edge cases | tie breaking logic does not work - see inline
Test implementations for `highest_score_from`| yes
**Player class** |
Uses an array to keep track of words played | yes
Uses existing code to create `highest_scoring_word` and `highest_word_score` | yes
Returns `true` or `false` from `won?` method | yes
Testing | some missing - see inline
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag | yes
`draw_tiles` method uses the parameter to draw a set of random tiles | yes
`tiles_remaining` method returns a count | yes
Tests `draw_tiles` edge cases | some missing

The big learning goals I'm looking to see demonstrated on this project are:

- Building complex logic (tie-breaking logic in `highest_score_from`)
- Test coverage
- Object composition (interaction between `Player` and `TileBag`)

You've made a good start on all of these, but there are substantial pieces missing for each. Please review my inline comments, and try and focus on these learning goals for Ride Share and next week's Hotel project. I'll be excited to see what you come up with!
